zebra-chain/src/types/amount.rs

```diff
@@ -0,0 +1,124 @@
+//! Module of types for working with validated zatoshi Amounts
+use std::{
+    convert::{TryFrom, TryInto},
+    marker::PhantomData,
+    ops::RangeInclusive,
+};
+
+/// A runtime validated type for representing amounts of zatoshis
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct Amount<C = NegativeAllowed>(i64, PhantomData<C>);
+
+impl<C> From<Amount<C>> for i64 {
+    fn from(amount: Amount<C>) -> Self {
+        amount.0
+    }
+}
+
+impl From<Amount<NonNegative>> for u64 {
+    fn from(amount: Amount<NonNegative>) -> Self {
+        amount.0 as _
+    }
+}
+
+impl<C> TryFrom<i64> for Amount<C>
+where
+    C: AmountConstraint,
+{
+    type Error = Error;
+
+    fn try_from(value: i64) -> Result<Self, Self::Error> {
+        C::validate(value).map(|v| Self(v, PhantomData))
+    }
+}
+
+impl<C> TryFrom<u64> for Amount<C>
+where
+    C: AmountConstraint,
+{
+    type Error = Error;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        let value = value
+            .try_into()
+            .map_err(|source| Error::Convert { value, source })?;
+
+        C::validate(value).map(|v| Self(v, PhantomData))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use proptest::prelude::*;
+    use std::fmt;
+
+    impl<C> Arbitrary for Amount<C>
+    where
+        C: AmountConstraint + fmt::Debug,
+    {
+        type Parameters = ();
+
+        fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+            C::valid_range().prop_map(|v| Self(v, PhantomData)).boxed()
+        }
+
+        type Strategy = BoxedStrategy<Self>;
+    }
+}
+
+#[derive(thiserror::Error, Debug, displaydoc::Display)]
+#[allow(missing_docs)]
+/// Errors that can be returned when validating `Amount`s
+pub enum Error {
+    /// input {value} is outside of valid range for zatoshi Amount, valid_range={range:?}
+    Contains {
+        range: RangeInclusive<i64>,
```

Comment:
@teor2345 I swapped these to `RangeInclusive` to hopefully prevent too many confusions over whether or not we should be including the bounds. Also I swapped to using `displaydoc` for the `Display` impl because this seemed like a good candidate for deduping docs and error messages.

Reply:
Thanks, I agree it's easier to understand when it is more explicit.